exchange/auction.go

```diff
@@ -60,111 +60,118 @@ func (a *auction) setRoundedPrices(priceGranularity openrtb_ext.PriceGranularity
+	var bids, vast, includeBidderKeys, includeWinners bool = targData.includeCacheBids, targData.includeCacheVast, targData.includeBidderKeys, targData.includeWinners
 	var errs []error
-	expectNumBids := valOrZero(bids, len(a.roundedPrices))
-	expectNumVast := valOrZero(vast, len(a.roundedPrices))
-	bidIndices := make(map[int]*openrtb.Bid, expectNumBids)
-	vastIndices := make(map[int]*openrtb.Bid, expectNumVast)
-	toCache := make([]prebid_cache_client.Cacheable, 0, expectNumBids+expectNumVast)
-	expByImp := make(map[string]int64)
-	competitiveExclusion := false
-	var hbCacheID string
-	if len(bidCategory) > 0 {
-		// assert:  category of winning bids never duplicated
-		if rawUuid, err := uuid.NewV4(); err == nil {
-			hbCacheID = rawUuid.String()
-			competitiveExclusion = true
-		} else {
-			errs = append(errs, errors.New("failed to create custom cache key"))
-		}
-	}
 
-	// Grab the imp TTLs
-	for _, imp := range bidRequest.Imp {
-		expByImp[imp.ID] = imp.Exp
-	}
-	for _, topBidsPerImp := range a.winningBidsByBidder {
-		for _, topBidPerBidder := range topBidsPerImp {
-			impID := topBidPerBidder.bid.ImpID
-			var customCacheKey string
-			var catDur string
-			useCustomCacheKey := false
-			if competitiveExclusion && topBidPerBidder == a.winningBids[impID] {
-				// set custom cache key for winning bid when competitive exclusion applies
-				catDur = bidCategory[topBidPerBidder.bid.ID]
-				if len(catDur) > 0 {
-					customCacheKey = fmt.Sprintf("%s_%s", catDur, hbCacheID)
-					useCustomCacheKey = true
-				}
+	if (bids || vast) && (includeBidderKeys || includeWinners) {
+		expectNumBids := valOrZero(bids, len(a.roundedPrices))
+		expectNumVast := valOrZero(vast, len(a.roundedPrices))
+		bidIndices := make(map[int]*openrtb.Bid, expectNumBids)
+		vastIndices := make(map[int]*openrtb.Bid, expectNumVast)
+		toCache := make([]prebid_cache_client.Cacheable, 0, expectNumBids+expectNumVast)
+		expByImp := make(map[string]int64)
+		competitiveExclusion := false
+		var hbCacheID string
+		if len(bidCategory) > 0 {
+			// assert:  category of winning bids never duplicated
+			if rawUuid, err := uuid.NewV4(); err == nil {
+				hbCacheID = rawUuid.String()
+				competitiveExclusion = true
+			} else {
+				errs = append(errs, errors.New("failed to create custom cache key"))
 			}
-			if bids {
-				if jsonBytes, err := json.Marshal(topBidPerBidder.bid); err == nil {
-					if useCustomCacheKey {
-						// not allowed if bids is true; log error and cache normally
-						errs = append(errs, errors.New("cannot use custom cache key for non-vast bids"))
+		}
+
+		// Grab the imp TTLs
+		for _, imp := range bidRequest.Imp {
+			expByImp[imp.ID] = imp.Exp
+		}
+		for ImpName, topBidsPerImp := range a.winningBidsByBidder {
+			//	winningBidsByBidder map[ImpName]map[openrtb_ext.BidderName]*pbsOrtbBid
+			for bidderName, topBidPerBidder := range topBidsPerImp {
+				impID := topBidPerBidder.bid.ImpID
```

Comment:
Above line can be eliminated if you replace **ImpName** on line 90 with **impID**.  Doing so undoes the change I made recently for https://jira.corp.appnexus.com/browse/PBS-348 based on the misperception that winningBidsByBidder is not indexed by impression ID.  See that ticket if you want further details.
If you don't remove this line as described, ImpName should be replaced by an underscore both because it isn't used otherwise and also because ImpName and the associated comment are a little misleading.

Reply:
Joe. It is probably best if I remove ImpName for the moment, don't you think? Maybe it'd be better to treat that as a separate issue.

Reply:
Argument in favor of changing ImpName to impID is that it is slightly more efficient:  impID is set only once instead of each time in each loop.  If you have any doubts about correctness, see the other references to winningBidsByBidder in auction.go and targeting.go, most notably setTargeting().

Reply:
Sure Joe. It'll get corrected in PBS-354